pkg/experiment/metastore/dlq/recovery.go

```diff
@@ -91,62 +91,71 @@ func (r *Recovery) recoverLoop(ctx context.Context) {
+			}
+		}
+	}()
+
+	b, err := r.readObject(ctx, path)
+	switch {
+	case err == nil:
+	case errors.Is(err, context.Canceled):
+		return err
+	case r.bucket.IsObjNotFoundErr(err):
+		// This is somewhat opportunistic: the error is likely caused by a competing recovery
+		// process that has already recovered the block, before we've discovered that the
+		// leadership has changed.
+		level.Warn(r.logger).Log("msg", "block metadata not found; skipping", "path", path)
 		return nil
+	default:
+		// This is somewhat opportunistic, as we don't know if the error is transient or not.
+		// we should consider an explicit retry mechanism with backoff and a limit on the
+		// number of attempts.
+		level.Warn(r.logger).Log("msg", "failed to read block metadata; to be retried", "err", err, "path", path)
+		return err
 	}
-	sshard := fields[1]
-	ulid := fields[3]
-	meta, err := r.get(ctx, metaPath)
-	if err != nil {
-		level.Error(r.logger).Log("msg", "failed to get block meta", "err", err, "path", metaPath)
+
+	var meta metastorev1.BlockMeta
+	if err = meta.UnmarshalVT(b); err != nil {
+		level.Error(r.logger).Log("msg", "invalid block metadata; skipping", "err", err, "path", path)
 		return nil
 	}
-	shard, _ := strconv.ParseUint(sshard, 10, 64)
-	if ulid != meta.Id || meta.Shard != uint32(shard) {
-		level.Error(r.logger).Log("msg", "unexpected block meta", "path", metaPath, "meta", fmt.Sprintf("%+v", meta))
+
+	switch _, err = r.metastore.AddRecoveredBlock(ctx, &metastorev1.AddBlockRequest{Block: &meta}); {
+	case err == nil:
 		return nil
-	}
-	if _, err = r.metastore.AddRecoveredBlock(ctx, &metastorev1.AddBlockRequest{Block: meta}); err != nil {
-		if raftnode.IsRaftLeadershipError(err) {
-			return err
-		}
-		level.Error(r.logger).Log("msg", "failed to add block", "err", err, "path", metaPath)
+	case status.Code(err) == codes.InvalidArgument:
+		level.Error(r.logger).Log("msg", "invalid block metadata", "err", err, "path", path)
 		return nil
+	case raftnode.IsRaftLeadershipError(err):
+		level.Warn(r.logger).Log("msg", "leadership change; recovery interrupted", "err", err, "path", path)
+		return err
+	default:
+		level.Error(r.logger).Log("msg", "failed to add block metadata; to be retried", "err", err, "path", path)
+		return err
 	}
-	err = r.bucket.Delete(ctx, metaPath)
-	if err != nil {
-		level.Error(r.logger).Log("msg", "failed to delete block meta", "err", err, "path", metaPath)
-	}
-	return nil
 }
 
-func (r *Recovery) get(ctx context.Context, metaPath string) (*metastorev1.BlockMeta, error) {
-	meta, err := r.bucket.Get(ctx, metaPath)
-	if err != nil {
-		return nil, err
-	}
-	metaBytes, err := io.ReadAll(meta)
-	if err != nil {
-		return nil, err
-	}
-	recovered := new(metastorev1.BlockMeta)
-	err = recovered.UnmarshalVT(metaBytes)
+func (r *Recovery) readObject(ctx context.Context, path string) ([]byte, error) {
+	rc, err := r.bucket.Get(ctx, path)
 	if err != nil {
 		return nil, err
 	}
-	return recovered, nil
```

Comment:
Here's where we leak the goroutine: the `bucket.Get` returns an `io.ReadClosed`, and it references the storage object client implementation until the `Close` call.